File: fft/Lua/mpd_display.lua

local lj_glfw = require "glfw"
local ffi = require "ffi"
local gl, glc, glu, glfw, glext = lj_glfw.libraries()
require 'luarocks.loader'
local fftw = require 'fftw3'

function read_buffer(data, stream)
    for x = 0, 1023 do
        stream:read(1)
        data[x] = (string.byte(stream:read(1)) + 128) % 256
    end
end

function draw_spectrogram(spectrogram)
    for i=1,512 do
        draw_spec_line(i-1, spectrogram)
    end
end

function draw_spec_line(x, spectrogram)
    gl.glBegin(glc.GL_QUAD_STRIP)
    for y=1,256 do
        colorval = spectrogram[(x*256)+y]
        colorval = 1 - colorval
        gl.glColor3f(colorval, colorval, colorval)
        local y_pos = y + 256
        gl.glVertex2f(x, y_pos)
        gl.glVertex2f(x+1, y_pos)
    end
    gl.glEnd()
end

function draw_line_scope(data)
    gl.glColor3d(1.0, 0.0, 0.0)
    gl.glBegin(glc.GL_LINE_STRIP)
    for i=1,1023 do
        gl.glVertex2f(i, data[i])
    end
    gl.glEnd()
end

function shift_spectrogram(spectrogram, new)
    for i=256+1,256*512 do
        spectrogram[i-256] = spectrogram[i]
    end
    for i=1, 256 do
        spectrogram[i+256*511] = new[i]
    end
end

function normalise(data)
    k = math.max(unpack(data, 2))
    if (0 < k) then
        for x=1,256 do
            data[x] = data[x]/k
        end
    end
    return data
end

function draw_rect(x, y)
    gl.glBegin(glc.GL_QUADS)
    y = y + 256
    x = x + 512
    gl.glVertex2f(x, y)
    gl.glVertex2f(x, 256)
    gl.glVertex2f(x+1, 256)
    gl.glVertex2f(x+1, y)

    gl.glEnd()
end

function draw_line_fft(power)
    for i = 1,256 do
        draw_rect(2*i-1, power[i]);
    end
end

function magnitude(out, draw_output)
    for i=1,512 do
        draw_output[i] = math.sqrt(out[i-1][0] * out[i-1][0]
                            + out[i-1][1] * out[i-1][1])/20.0
    end
end

lj_glfw.init()
local window = lj_glfw.Window(1024, 512, "MPD Visualizer")

window:setKeyCallback(
function(window, key)
    if key == string.byte('Q') then
        window:setShouldClose(glc.GL_TRUE)
    end
end)
window:makeContextCurrent()

gl.glOrtho(0, 1024, 0, 512, 0, 1);
gl.glClearColor(1.0, 1.0, 1.0, 0);

stream = io.open("/tmp/mpd.fifo", "rb")

local data = ffi.new("double [1024]", 1024)
local fft_cplx_out = ffi.new("double [1024][2]")
local p = fftw.plan_dft_r2c_1d(1024, data, fft_cplx_out, 0)
local spectrogram = {}
local fft_draw_out = {}

for i=1,256*512+1 do
    spectrogram[i] = 0
end

while not window:shouldClose() do
    gl.glClear(glc.GL_COLOR_BUFFER_BIT)

    read_buffer(data, stream)
    fftw.execute(p)

    magnitude(fft_cplx_out, fft_draw_out)

    draw_line_scope(data)
    draw_line_fft(fft_draw_out)

    shift_spectrogram(spectrogram, normalise(fft_draw_out))
    draw_spectrogram(spectrogram)

    lj_glfw.pollEvents()
    window:swapBuffers()
end

io.close(stream)
window:destroy()
lj_glfw.terminate()

File: fft/Haskell/mpd-display.hs

module Main where
import qualified Graphics.UI.GLFW as GLFW
import Graphics.Rendering.OpenGL.Raw

import System.Exit (exitSuccess)
import Control.Monad (zipWithM_, unless)

import System.IO
import Data.ByteString (hGet, unpack)
import Data.Word

import Data.Sequence
import Math.FFT.Base (dftRC)
import Data.Array.CArray (array, elems)
import Data.Array.CArray.Base
import Data.Complex
import Data.Foldable (toList)

everySecond :: [a] -> [a]
everySecond (_:a:rest) = a : everySecond rest
everySecond _ = []

scale :: [Word8] -> [Word8]
scale = map (\x -> (x + 128) `mod` 255)

initGL :: GLFW.Window -> IO ()
initGL _ = do
  glHint gl_DONT_CARE gl_FASTEST
  glClearColor 1.0 1.0 1.0 0.0
  glOrtho 0.0 1024.0 0.0 512.0 0.0 1.0

shutdown :: Handle -> GLFW.Window -> IO ()
shutdown stream win = do
  GLFW.destroyWindow win
  GLFW.terminate
  hClose stream
  putStrLn "Program quit."
  exitSuccess
  return ()

generateSpectrogram :: Seq [Double]
generateSpectrogram =
  Data.Sequence.replicate 512 $ Prelude.replicate 256 0

drawLineSignal :: [Double] -> IO ()
drawLineSignal signal = do
    glBegin gl_LINE_STRIP
    zipWithM_ glVertex2f [1.0 .. 1024.0] $ map realToFrac signal
    glEnd

drawFFT :: [Double] -> IO ()
drawFFT fftSignal = do
  let drawRect x y = do
        glBegin gl_QUADS
        glColor3f 1 0 0
        let xnew = 2 * x + 512
        let ynew = y + 256
        glVertex2f xnew ynew
        glVertex2f xnew 256
        glVertex2f (xnew + 1) 256
        glVertex2f (xnew + 1) ynew
        glEnd

  zipWithM_ (\x y -> drawRect x $ min 256 y)
    [1 .. 256]
    $ map (\x -> realToFrac x / 20) fftSignal

drawSpectrogram :: Seq [Double] -> IO ()
drawSpectrogram spectrogram = do
  let drawSpecLine x l =
        let normalize line =
              let m = 0.0625 * maximum line in
              if m > 2 then map (/ m) line else line in
         let line = normalize l in do
          glBegin gl_QUAD_STRIP
          zipWithM_ (\y intensity ->
                      let colorval = 1 - realToFrac intensity
                          ypos = y + 256 in do
                       glColor3f colorval colorval colorval
                       glVertex2f x ypos
                       glVertex2f (x + 1) ypos)
                    [1 .. 256]
                    line
          glEnd
  zipWithM_ drawSpecLine [1 .. 512] $ toList spectrogram

keyPressed :: GLFW.KeyCallback
keyPressed win GLFW.Key'Q _ GLFW.KeyState'Pressed _ = GLFW.setWindowShouldClose win True
keyPressed _   _          _ _                     _ = return ()

mainDrawLoop :: Seq [Double] -> Handle -> GLFW.Window -> IO ()
mainDrawLoop spectrogram stream win = do
  shouldClose <- GLFW.windowShouldClose win
  unless shouldClose $ do
    glClear $ fromIntegral gl_COLOR_BUFFER_BIT
    glColor3f 1 0 0
    contents <- hGet stream 2048
    let signal = map fromIntegral . scale $ everySecond $ unpack contents
    let fftIn = array (1, 1024) $ Prelude.zip [1 .. 1024] signal :: CArray Int Double
    let fftSignal = map (realPart . abs) $ elems $ dftRC fftIn
    let newSpectrogram = case viewl spectrogram of
                              _ :< end -> (|>) end fftSignal

    drawLineSignal signal
    drawFFT fftSignal
    drawSpectrogram newSpectrogram

    GLFW.pollEvents
    GLFW.swapBuffers win
    mainDrawLoop newSpectrogram stream win

main :: IO ()
main = do
     True <- GLFW.init
     Just win <- GLFW.createWindow 1024 512 "Haskell MPD Visuaizer"
                 Nothing Nothing
     GLFW.makeContextCurrent (Just win)

     stream <- openFile "/tmp/mpd.fifo" ReadMode
     GLFW.setKeyCallback win (Just keyPressed)

     initGL win
     let s = generateSpectrogram
     mainDrawLoop s stream win

     shutdown stream win
